### PR TITLE
Add key repetition for the wayland backend

### PR DESCRIPTION
**Pull request by trimental**
- [x] Tested on all platforms changed
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users

This PR adds the feature of key repetition to the wayland backend. This PR is requested prematurely so that it may be reviewed before the release of smithay-client-toolkit 0.3.0, currently it points to the git version as a dependency.

I have tested the changes and added a small update to the changelog (might need more explanation and better wording?), however I cannot find any documentation of key repetition (I assume this is just standard behavior).

Please look to https://github.com/Smithay/client-toolkit/pull/16 for further information about the new `map_keyboard_with_repeat()` function

### CODE REVIEW

**Comment by vberger** [comment]
This is a simpler implementation of key repetition, based on a thread.

It has been developed in #371 that no background thread should be needed. I do have some plans to introduce some real, proper event loop capability to my wayland libs (in a way making it possible to have good support for secondary event sources, including timers). But I don't know when I'll have the bandwidth to do this (this is related to https://github.com/Smithay/wayland-rs/issues/185 for those interested).

In the meantime, this works, and leaves room to easily implement stuff like disabling or discriminating key repetition events, depending on what is decided wrt #310.

File: Cargo.toml [hunk 1]
```diff
@@ -55,7 +55,7 @@ features = [
 
 [target.'cfg(any(target_os = "linux", target_os = "dragonfly", target_os = "freebsd", target_os = "openbsd", target_os = "netbsd"))'.dependencies]
 wayland-client = { version = "0.20.10", features = [ "dlopen", "egl", "cursor"] }
-smithay-client-toolkit = "0.2.6"
+smithay-client-toolkit = { git = "https://github.com/Smithay/client-toolkit" }
```
Comment (SirCmpwn) [inline Cargo.toml]: You can use 0.3.0 now.

Comment (trimental) [reply Cargo.toml]: I'm going to change this pull request to 'Upgrade to smithay-client-toolkit 0.3.0' as 0.3.0 contains few breaking changes from https://github.com/Smithay/client-toolkit/pull/25

Comment (trimental) [reply Cargo.toml]: Sorry my bad, it seems none of those breaking changes apply

**Comment by francesca64** [comment]
Oh, that would be fine. The checklist system isn't perfect, though I don't really know if there's a better way... anyway, is this PR ready for merge?

File: src/platform/linux/wayland/keyboard.rs [hunk 2]
```diff
@@ -1,77 +1,110 @@
...
 
 pub fn init_keyboard(
     keyboard: NewProxy<wl_keyboard::WlKeyboard>,
     sink: Arc<Mutex<EventsLoopSink>>,
+    events_loop_proxy: EventsLoopProxy,
 ) -> Proxy<wl_keyboard::WlKeyboard> {
-    // { variables to be captured by the closure
-    let mut target = None;
+    // { variables to be captured by the closures
+    let target = Arc::new(Mutex::new(None));
```
Comment (francesca64) [inline src/platform/linux/wayland/keyboard.rs]: `Default::default()` should be equivalent to this.

**Comment by trimental** [comment]
I believe it should be in a state to merge after implementing your review point. I just want to be sure though that my understanding is correct that there is no difference between key press and key repeat events in winit and that they should be indistinguishable.

File: src/platform/linux/wayland/keyboard.rs (hunk 2 above)
Comment (francesca64) [reply src/platform/linux/wayland/keyboard.rs]: Oh, since the type isn't inferred in this case, then it's preferable the way it already is. Sorry!

**Comment by francesca64** [comment]
That's correct; our API presently has no indication of a key press being a repeat.

**Comment by SirCmpwn** [comment]
Thanks! I'm sure you guys have your own release cycle but I wanted to mention that I'm looking forward to seeing this tagged for alacritty's sake.

**Comment by francesca64** [comment]
This is going to be included in a new release as part of #630, which will merge as soon as the CI passes!

**Comment by SirCmpwn** [comment]
:tada:

**Pull request merged**